openml/extensions/onnx/extension.py

```diff
@@ -0,0 +1,844 @@
+import importlib
+import json
+import logging
+import re
+import sys
+import time
+import warnings
+import zlib
+import math
+from collections import OrderedDict  # noqa: F401
+from distutils.version import LooseVersion
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import onnx
+import numpy as np
+import pandas as pd
+import scipy.sparse
+import mxnet as mx
+import mxnet.contrib.onnx as onnx_mxnet
+from mxnet import nd, gluon, autograd
+from google.protobuf import json_format
+
+import openml
+from openml.exceptions import PyOpenMLError
+from openml.extensions import Extension, register_extension
+from openml.flows import OpenMLFlow
+from openml.runs.trace import OpenMLRunTrace, OpenMLTraceIteration
+from openml.tasks import (
+    OpenMLTask,
+    OpenMLSupervisedTask,
+    OpenMLClassificationTask,
+    OpenMLRegressionTask,
+)
+
+from .config import criterion_gen, optimizer, batch_size, epoch_count, sanitize_value, context
+
+if sys.version_info >= (3, 5):
+    from json.decoder import JSONDecodeError
+else:
+    JSONDecodeError = ValueError
+
+DEPENDENCIES_PATTERN = re.compile(
+    r'^(?P<name>[\w\-]+)((?P<operation>==|>=|>)'
+    r'(?P<version>(\d+\.)?(\d+\.)?(\d+)?(dev)?[0-9]*))?$'
+)
+
+SIMPLE_NUMPY_TYPES = [nptype for type_cat, nptypes in np.sctypes.items()
+                      for nptype in nptypes if type_cat != 'others']
+SIMPLE_TYPES = tuple([bool, int, float, str] + SIMPLE_NUMPY_TYPES)
+
+
+LAYER_PATTERN = re.compile(r'layer\d+\_(.*)')
+
+
+class OnnxExtension(Extension):
+    """Connect ONNX to OpenML-Python."""
+
+    ################################################################################################
+    # General setup
+
+    @classmethod
+    def can_handle_flow(cls, flow: 'OpenMLFlow') -> bool:
+        """Check whether a given flow describes an ONNX specification.
+
+        This is done by parsing the ``external_version`` field.
+
+        Parameters
+        ----------
+        flow : OpenMLFlow
+
+        Returns
+        -------
+        bool
+        """
+        return cls._is_onnx_flow(flow)
+
+    @classmethod
+    def can_handle_model(cls, model: Any) -> bool:
+        """Check whether a model is an instance of ``onnx.ModelProto``.
+
+        Parameters
+        ----------
+        model : Any
+
+        Returns
+        -------
+        bool
+        """
+        return isinstance(model, onnx.ModelProto)
+
+    ################################################################################################
+    # Methods for flow serialization and de-serialization
+
+    def flow_to_model(self, flow: 'OpenMLFlow', initialize_with_defaults: bool = False) -> Any:
+        """Initializes an ONNX model representation based on a flow.
+
+        Parameters
+        ----------
+        flow : mixed
```

Comment:
The flow parameter seems to have the OpenMLFlow type, why does it say it's mixed here?